src/test/scala/chiselTests/ReinterpretCast.scala

```diff
@@ -0,0 +1,68 @@
+// See LICENSE for license details.
+
+package chiselTests
+
+import org.scalatest._
+
+import chisel3._
+import chisel3.experimental.FixedPoint
+import chisel3.testers.BasicTester
+import chisel3.util._
+import chisel3.core.DataMirror
+
+class AsBundleTester extends BasicTester {
+  class MultiTypeBundle extends Bundle {
+    val u  = UInt(4.W)
+    val s  = SInt(4.W)
+    val fp = FixedPoint(4.W, 3.BP)
+  }
+
+  val rawData = ((4 << 8) + (15 << 4) + (12 << 0)).U
+  val bunFromBits = rawData.asTypeOf(new MultiTypeBundle)
+
+  assert(bunFromBits.u === 4.U)
+  assert(bunFromBits.s === -1.S)
+  assert(bunFromBits.fp === FixedPoint.fromDouble(-0.5, width=4, binaryPoint=3))
+
+  stop()
+}
+
+class AsVecTester extends BasicTester {
+  val rawData = ((15 << 12) + (0 << 8) + (1 << 4) + (2 << 0)).U
+  val vec = rawData.asTypeOf(Vec(4, SInt(4.W)))
+
+  assert(vec(0) === 2.S)
+  assert(vec(1) === 1.S)
+  assert(vec(2) === 0.S)
+  assert(vec(3) === -1.S)
+
+  stop()
+}
+
+class AsBitsTruncationTester extends BasicTester {
+  val truncate = (64 + 3).U.asTypeOf(UInt(3.W))
+  val expand   = 1.U.asTypeOf(UInt(3.W))
+
+  assert( DataMirror.widthOf(truncate).get == 3 )
+  assert( truncate === 3.U )
+  assert( DataMirror.widthOf(expand).get == 3 )
+  assert( expand === 1.U )
```

Comment:
I think there's a good argument for an explicit reflection-like API, especially since there are subtleties with width (inferred width doesn't get computed until after Chisel).

But in any case we should standardize on one, and deprecate the other.